audbackend/core/artifactory.py

```diff
@@ -105,30 +124,65 @@ def _ls(
         )
         return [p.name for p in audfactory.path(path)]
 
+    def _path(
+            self,
+            path: str,
+            version: typing.Optional[str],
+            ext: str,
+    ) -> str:
+        r"""Convert to backend path.
+
+        <folder>/<name>.<ext>
+        ->
+        <host>/<repository>/<folder>/<name>/<version>/<name>-<version>.<ext>
+
+        """
+        folder = self._folder(path, ext)
+        name = os.path.basename(folder)
+        path = f'{folder}/{version}/{name}-{version}{ext}'
+
+        return path
+
     def _put_file(
             self,
             src_path: str,
             dst_path: str,
+            version: str,
+            ext: str,
             verbose: bool,
     ):
         r"""Put file to backend."""
+        dst_path = self._path(dst_path, version, ext)
         audfactory.deploy(src_path, dst_path, verbose=verbose)
 
     def _remove_file(
             self,
             path: str,
+            version: str,
+            ext: str,
     ):
         r"""Remove file from backend."""
+        path = self._path(path, version, ext)
         audfactory.path(path).unlink()
 
     def _versions(
             self,
-            folder: str,
-            name: str,
+            path: str,
+            ext: str,
     ) -> typing.List[str]:
         r"""Versions of a file."""
-        group_id = audfactory.path_to_group_id(folder)
-        return audfactory.versions(self.host, self.repository, group_id, name)
+        folder = self._folder(path, ext)
+        folder = audfactory.path(folder)
+
+        try:
+            vs = [os.path.basename(str(f)) for f in folder if f.is_dir]
+        except (FileNotFoundError, RuntimeError):
+            vs = []
+
+        # filter out versions of files with different extension
+        vs = [v for v in vs if self._exists(path, v, ext)]
```

Comment:
I remember what the issue might be here. `audfactory.versions()` does not really handle this case as Artifactory was not designed to store the same file with different extensions (for whatever reasons).

As you now filter the versions here anyway could we stay with `audfactory.versions()` in the line before or was there another reason you replaced it?

Reply:
Good point, it should be relatively easy to deprecate `audfactory` as it is most likely only needed inside `audbackend`. The main thing we would need to add to `audbackend` would be the code of `audfactory.authentification()`.

So let's stay here without using `audfactory.versions()`.

Reply:
I created https://github.com/audeering/audbackend/issues/50